content/post/new-python-api.md

```diff
@@ -0,0 +1,263 @@
+        return bs.decode("utf-8")
+```
+
+The only part of the decoder that really needs to have any logic is the part that takes bytes and returns a message; the first item can be described by an integer and the second one can be described by a string that is passed as an argument to `struct.unpack(...)` to tell it how to turn the incoming bytes into a number. With these changes, a decoder now looks like this:
+```python
+@wallaroo.decoder(header_length=4, length_fmt=">I")
+def decoder(bs):
+    return bs.decode("utf-8")
+```
+
+## Why We Created a New API
+
+Before we had even published the blog post about the original Python API, we had discussed ideas for things that we could do to improve the API. One of the ideas that we had considered was using decorators to cut down on the number of classes and thus reduce the amount of code that needed to be written. When we asked for feedback, several people suggested using decorators to improve the API, so we felt that our earlier idea had been validated. We went ahead and designed a new decorator-based API.
+
+## The New Python API
+
+### A Motivating Example
+
+We will start with the canonical streaming data processing application, Word Count. A stream of input text is analyzed and the total number of times each word has been seen is reported. [The example](https://github.com/WallarooLabs/wallaroo/blob/release/examples/python/word_count/word_count.py) in it's entirety is in our [GitHub repository](https://github.com/WallarooLabs/wallaroo/tree/release).
+
+We will make the following assumptions:
+* Incoming messages will come from a TCP connection and be sent to
+  another TCP connection.
+* Words are sent to the system in messages that can contain zero or
+  more words.
+* Incoming messages consist of a string.
+* Outgoing messages consist of a word and the number of times that
+  word has been seen in the event stream.
+
+![Word Count Diagram](/images/post/python-api/word-count-diagram.png)
+
+In our example, we will also split the state (the number of times each word has been seen) into 26 partitions, where each partition handles words that start with different letters. For example "acorn" and "among" would go to the "a" partition, while "bacon" would go to the "b" partition.
+
+This application will process messages as they arrive. This contrasts with some other streaming data processing systems that are designed around processing messages in micro-batches. This results in lower latencies because message processing is not delayed.
+
+### Wallaroo's Core Abstractions
+
+In order to understand the Python API, it is important to understand Wallaroo's core abstractions:
+
+* State -- Accumulated result of data stored over the course of time.
+* Computation -- Code that transforms an input to an
+  output.
+* State Computation -- Code that takes an input and a state
+  object, operates on that input and state
+  (possibly making state updates), and optionally produces an output.
```

Comment:
We use `state` and `state object` here and then later we also use `state partition` and `state entity`. It seems there is at least some ambiguity between `state object` and `state entity`. Maybe they're supposed to be the same thing?

Reply:
Changing "state object" to "state entity".